pkg/providers/snow/snow_test.go

```diff
@@ -0,0 +1,200 @@
+	cluster          *v1alpha1.Cluster
+	clusterSpec      *cluster.Spec
+	datacenterConfig *v1alpha1.SnowDatacenterConfig
+	machineConfigs   map[string]*v1alpha1.SnowMachineConfig
+}
+
+func givenFullAPIObjects(t *testing.T, fileName string) apiObjects {
+	clusterSpec := test.NewFullClusterSpec(t, path.Join(testDataDir, fileName))
+	datacenterConfig, err := v1alpha1.GetSnowDatacenterConfig(path.Join(testDataDir, fileName))
+	if err != nil {
+		t.Fatalf("unable to get snow datacenter config from file: %v", err)
+	}
+	machineConfigs, err := v1alpha1.GetSnowMachineConfigs(path.Join(testDataDir, fileName))
+	if err != nil {
+		t.Fatalf("unable to get snow machine configs from file")
+	}
+
+	return apiObjects{
+		cluster:          clusterSpec.Cluster,
+		clusterSpec:      clusterSpec,
+		datacenterConfig: datacenterConfig,
+		machineConfigs:   machineConfigs,
+	}
+}
+
+func givenProvider(t *testing.T) *snowProvider {
+	return newProvider(t, nil, nil, nil, nil)
+}
+
+func givenEmptyClusterSpec() *cluster.Spec {
+	return test.NewClusterSpec(func(s *cluster.Spec) {
+		s.VersionsBundle.KubeVersion = "1.21"
+	})
+}
+
+func newProvider(t *testing.T, datacenterConfig *v1alpha1.SnowDatacenterConfig, machineConfigs map[string]*v1alpha1.SnowMachineConfig, clusterConfig *v1alpha1.Cluster, kubectl ProviderKubectlClient) *snowProvider {
+	return NewProvider(
+		datacenterConfig,
+		machineConfigs,
+		clusterConfig,
+		kubectl,
+		nil,
+		test.FakeNow,
+	)
+}
+
+func setupContext(t *testing.T, envVar string) {
+	envVarOrgValue, isSet := os.LookupEnv(envVar)
+	t.Cleanup(func() {
+		if isSet {
+			os.Setenv(envVar, envVarOrgValue)
+		} else {
+			os.Unsetenv(envVar)
+		}
+	})
+}
+
+func TestName(t *testing.T) {
+	tt := newSnowTest(t)
+	tt.Expect(tt.provider.Name()).To(Equal(expectedSnowProviderName))
+}
+
+func TestSetupAndValidateCreateClusterSuccess(t *testing.T) {
+	tt := newSnowTest(t)
+	setupContext(t, credsFileEnvVar)
+	setupContext(t, certsFileEnvVar)
+	os.Setenv(credsFileEnvVar, credsFilePath)
+	os.Setenv(certsFileEnvVar, certsFilePath)
+	err := tt.provider.SetupAndValidateCreateCluster(tt.ctx, tt.clusterSpec)
+	tt.Expect(err).To(Succeed())
+}
+
+func TestSetupAndValidateCreateClusterNoCredsEnv(t *testing.T) {
+	tt := newSnowTest(t)
+	setupContext(t, credsFileEnvVar)
+	setupContext(t, certsFileEnvVar)
+	os.Setenv(certsFileEnvVar, certsFilePath)
+	err := tt.provider.SetupAndValidateCreateCluster(tt.ctx, tt.clusterSpec)
+	tt.Expect(err).To(MatchError(ContainSubstring("EKSA_SNOW_DEVICES_CREDENTIALS_FILE is not set or is empty")))
+}
+
+func TestSetupAndValidateCreateClusterNoCertsEnv(t *testing.T) {
+	tt := newSnowTest(t)
+	setupContext(t, credsFileEnvVar)
+	os.Setenv(credsFileEnvVar, credsFilePath)
+	err := tt.provider.SetupAndValidateCreateCluster(tt.ctx, tt.clusterSpec)
+	tt.Expect(err).To(MatchError(ContainSubstring("EKSA_SNOW_DEVICES_CA_BUNDLES_FILE is not set or is empty")))
+}
+
+// TODO: add more tests (multi worker node groups, unstacked etcd, etc.)
```

Comment:
Suggestion: maybe keep this method tests at a minimum and just test the methods that return api structs? That avoids having to create a ton of yaml files and probably makes tests more "unitary"

Reply:
not sure if i 100% agree. this method returns cp and md yaml templates in []byte. it makes sense to test against yamls. 
also the api struct test is covered in the two apibuilder_test